activerecord/lib/active_record/delay_touching/state.rb

```diff
@@ -0,0 +1,71 @@
+module ActiveRecord
+  # = Active Record Delay Touching
+  module DelayTouching
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
+      def initialize
+        @records = Hash.new { SortedSet.new }
+        @already_updated_records = Hash.new { SortedSet.new }
+      end
+
+      def updated(klass, attrs, records)
+        @already_updated_records[[klass, attrs]] += records
+      end
+
+      # Return the records grouped by class and columns that were touched:
+      #
+      #   {
+      #     [Owner, [:updated_at]]               => SortedSet.new([owner1, owner2]),
+      #     [Pet,   [:neutered_at, :updated_at]] => SortedSet.new([pet1]),
+      #     [Pet,   [:updated_at]]               => SortedSet.new([pet2])
+      #   }
+      #
+      # As a side-effect, clears out the list of records.
+      def get_and_clear_records
+        records = @records
+        @records = Hash.new { SortedSet.new }
+        records
+      end
+
+      def more_records?
+        @records.present?
+      end
+
+      def add_record(record, columns)
```

Comment:
How about this :smile:
```ruby
def add_record(record, extra_columns)
  key = [
    record.class,
    record.timestamp_attributes_for_update_in_model.concat(extra_columns).map(&:to_sym).sort
  ]

  @records[key] += [record] unless @already_updated_records[key].include?(record)
end
```

`timestamp_attributes_for_update_in_model` uses `select` which returns an empty array if there's nothing to select. This means we can count on the return value responding to `concat` and just use that regardless.

Also it bothers me a bit the value for the key is wrapped in an array `[record]`. Doesn't `SortedSet`s have a method to add a single item without needing it to be an array or enumerable?